// @netlify/sdk--ui-core/dist/message_schema/mocks.js
import fc from "fast-check";
import { Surfaces } from "../surfaces.js";
import { parseNonUndefined } from "../utils/parsers.js";
import { UI_EXTENSION_HOST_INITIALIZE_EVENT_TYPE, } from "./main.js";
const hexa = () => {
    const dict = "0123456789abcdef";
    return fc.integer({ min: 0, max: 15 }).map((n) => {
        const value = dict.at(n);
        if (value === undefined) {
            throw new RangeError("no item in dictionary available, programmer error");
        }
        return value;
    });
};
const hexaString = (constraints = {}) => {
    return fc.string(Object.assign(Object.assign({}, constraints), { unit: hexa() }));
};
export const SurfaceArbitrary = fc.constantFrom(...Object.values(Surfaces));
export const UIExtensionHostInitializeEventDataContextAuthArbitrary = fc.record({
    netlifyToken: hexaString({ minLength: 1 }),
    providerToken: hexaString({ minLength: 1 }),
    providerTokens: fc.array(fc.record({
        externalId: hexaString({ minLength: 1 }),
        token: hexaString({ minLength: 1 }),
        connectionId: hexaString({ minLength: 1 }),
    }), { minLength: 0 }),
});
export const UIExtensionHostInitializeEventDataContextUserArbitrary = fc.record({
    id: hexaString({ minLength: 1 }),
});
export const UIExtensionHostInitializeEventDataContextVisualEditorArbitrary = fc.record({
    webhookBaseURL: hexaString({ minLength: 1 }),
    projectId: hexaString({ minLength: 1 }),
    notifyVisualEditor: fc.boolean(),
});
export const UIExtensionHostInitializeEventDataContextArbitrary = fc.record({
    auth: UIExtensionHostInitializeEventDataContextAuthArbitrary,
    user: UIExtensionHostInitializeEventDataContextUserArbitrary,
    visualEditor: UIExtensionHostInitializeEventDataContextVisualEditorArbitrary,
    configurationId: hexaString({ minLength: 1 }),
    dataLayerId: hexaString({ minLength: 1 }),
    projectId: hexaString({ minLength: 1 }),
    siteId: hexaString({ minLength: 1 }),
    teamId: hexaString({ minLength: 1 }),
    accountId: hexaString({ minLength: 1 }),
    theme: fc.constantFrom("light", "dark"),
    deployId: hexaString({ minLength: 1 }),
});
export const UIExtensionHostInitializeEventDataExtensionArbitrary = fc.record({
    id: hexaString({ minLength: 1 }),
    name: hexaString({ minLength: 1 }),
    slug: hexaString({ minLength: 1 }),
});
export const UIExtensionHostInitializeEventDataSurfaceArbitrary = fc.record({
    name: SurfaceArbitrary,
});
export const UIExtensionHostInitializeEventDataArbitrary = fc.record({
    context: UIExtensionHostInitializeEventDataContextArbitrary,
    extension: UIExtensionHostInitializeEventDataExtensionArbitrary,
    surface: UIExtensionHostInitializeEventDataSurfaceArbitrary,
});
export const UIExtensionHostInitializeEventArbitrary = fc.record({
    type: fc.constant(UI_EXTENSION_HOST_INITIALIZE_EVENT_TYPE),
    data: UIExtensionHostInitializeEventDataArbitrary,
});
export const createMockUIExtensionHostInitializeEvent = (overrides = {}) => {
    return Object.assign(Object.assign({}, parseNonUndefined(fc.sample(UIExtensionHostInitializeEventArbitrary, 1)[0])), overrides);
};
