// @netlify/sdk--ui-react/dist/components/Table/Table.js
import { jsx, jsxs } from "react/jsx-runtime";
import { cn } from "../classnames.js";
const sharedCellStyles = [
  "tw-px-4",
  "tw-py-3",
  "@lg:first:tw-pl-6 supports-not-container-queries:md:first:tw-pl-6",
  "@lg:last:tw-pr-6 supports-not-container-queries:md:last:tw-pr-6",
  "empty:!tw-p-0 last:empty:!tw-pr-2 empty:tw-w-0",
  "tw-text-left"
];
const TableHead = ({
  children,
  className,
  ...otherProps
}) => /* @__PURE__ */ jsx("thead", { ...otherProps, children });
const TableHeader = ({
  children,
  className,
  ...otherProps
}) => {
  return /* @__PURE__ */ jsx(
    "th",
    {
      scope: "col",
      className: cn(...sharedCellStyles, "tw-text-muted", className),
      ...otherProps,
      children
    }
  );
};
const TableRow = ({
  children,
  className,
  ...otherProps
}) => {
  return /* @__PURE__ */ jsx("tr", { className: cn("even:tw-bg-bg", className), ...otherProps, children });
};
const TableData = ({
  children,
  className,
  ...otherProps
}) => /* @__PURE__ */ jsx(
  "td",
  {
    className: cn(
      ...sharedCellStyles,
      "tw-tabular-nums",
      "tw-text-text",
      className
    ),
    ...otherProps,
    children
  }
);
const TableBody = ({
  children,
  className,
  ...otherProps
}) => /* @__PURE__ */ jsxs("tbody", { ...otherProps, children: [
  " ",
  children,
  " "
] });
const Table = ({ children, className, ...otherProps }) => /* @__PURE__ */ jsx(
  "table",
  {
    className: cn(
      "tw-@container",
      "tw-w-full",
      "tw-border-collapse",
      "tw-transition-opacity",
      "tw-duration-200",
      className
    ),
    ...otherProps,
    children
  }
);
export {
  Table,
  TableBody,
  TableData,
  TableHead,
  TableHeader,
  TableRow
};
